src/Lykke.Cqrs/Middleware/CommandInterceptorsQueue.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Lykke.Cqrs.Abstractions.Middleware;
+
+namespace Lykke.Cqrs.Middleware
+{
+    internal class CommandInterceptorsQueue
+    {
+        private readonly List<ICommandInterceptor> _commandInterceptors = new List<ICommandInterceptor>();
+
+        internal void AddInterceptor(ICommandInterceptor commandInterceptor)
+        {
+            _commandInterceptors.Add(commandInterceptor);
+        }
+
+        internal Task<CommandHandlingResult> RunInterceptorsAsync(
+            object command,
+            object handlerObject,
+            IEventPublisher eventPublisher,
+            Func<object, IEventPublisher, CommandHandlingResult> actualHandler)
+        {
+            var actualHandlerInterceptor = new CommandActualHandlerInterceptor(actualHandler);
+            var interceptor = _commandInterceptors.FirstOrDefault() ?? actualHandlerInterceptor;
+
+            var commonContext = new CommandInterceptionContext
+            {
+                Command = command,
+                HandlerObject = handlerObject,
+                EventPublisher = eventPublisher,
+            };
+            var interceptorContext = new CommandInterceptorContext(
+                commonContext,
+                0,
+                this,
+                actualHandlerInterceptor);
+
+            return interceptor.InterceptAsync(interceptorContext);
+        }
+
+        internal ICommandInterceptor ResolveNext(int currentInterceptorIndex, ICommandInterceptor actualHandlerInterceptor)
```

Comment:
Looks like actualHandlerInterceptor is not needed here. This method can just return when all interceptors in the queue are already iterated. Calling method can decide what to do in such case. And after this modification CommandActualHandlerInterceptor will be unnecessary, you will be able just pass Func to the context.

And since this method can return null, I suggest to rename it to the TryResolveNext